### ISSUE TITLE
OpenCL.CLException

### ISSUE THREAD

**Issue by fcogneaux** [m4rs-mt/ILGPU#1169]
### Question

Hi m4rs-mt and team, kudos for this library!

I'm facing an issue trying to migrate a resampling computation from CPU to GPU.

I simplified the code below as much as I could to reproduce the issue so it may not make much sens but, basically, it's a correlation computation nested in a resampling (hence the double loop in the kernel).

The issue: as soon as the size of the data exceeds a given number (8000   rows * 1000 samples on my development laptop) I get this error:

ILGPU.Runtime.OpenCL.CLException
  HResult=0x80131500
  Message=Exception of type 'ILGPU.Runtime.OpenCL.CLException' was thrown.
  Source=ILGPU
  StackTrace:
   at ILGPU.Runtime.OpenCL.CLAccelerator.SynchronizeInternal() in /_/Src/ILGPU/Runtime/OpenCL/CLAccelerator.cs:line 388

Unfortunately, the inner exception is "null", leaving me without a clue. It works like a charm with smaller values or with larger values but using the CPUAccelerator, thus it doesn't seem related to a boundaries index issue and I don't know how to debug... It shouldn't be related to memory size either, as all the data fit into about 40Mo, while the free memory GPU exceeds 7Go.

Thanks a lot for your help (I'm new at GPU programming, I apologize in advance if I'm wasting your time with an beginner question...) !

The code:

```
using ILGPU;
using ILGPU.Runtime;

namespace Test {
    internal class Test {

        const int rows = 8000;       // Fails from 8000 on my laptop...
        const int sample = 1000;
        const int resample = 25000;

        static void Main(string[] args) {

            Random oRnd = new();

            // Context and accelerator

            using Context oContext = Context.CreateDefault();
            using Accelerator oAccelerator = oContext.GetPreferredDevice(preferCPU: false).CreateAccelerator(oContext);

            Console.WriteLine($"Start ({oAccelerator.Name})...");

            // Data preparation

            // The data (random here)
            float[] data = new float[rows * sample];
            for (int i = 0; i < data.Length; i++) {
                data[i] = oRnd.NextSingle();
            }

            // Some pre-computations
            float[] opt = new float[sample];
            for (float i = 1; i < opt.Length; i++) {
                opt[(int)i] = 1 / ((i + 1) * i);
            }

            // The results
            float[] results = new float[resample];

            // Create the memory buffers
            using MemoryBuffer1D<float, Stride1D.Dense> mOpt = oAccelerator.Allocate1D<float>(opt);
            using MemoryBuffer1D<float, Stride1D.Dense> mData = oAccelerator.Allocate1D<float>(data);
            using MemoryBuffer1D<float, Stride1D.Dense> mResult = oAccelerator.Allocate1D<float>(results);

            // Load / precompile the kernel
            Action<Index1D, ArrayView<float>, ArrayView<float>, ArrayView<float>> loadedKernel =
                oAccelerator.LoadAutoGroupedStreamKernel<Index1D, ArrayView<float>, ArrayView<float>, ArrayView<float>>(Kernel);

            // Start the kernel
            loadedKernel(resample, mOpt.View, mData.View, mResult.View);

            // Synchronize
            oAccelerator.Synchronize();     // <- exception here

            // Donwload the results
            results = mResult.GetAsArray1D();

            Console.WriteLine($"Done!");
        }

        static void Kernel(Index1D nIndex, ArrayView<float> aOpt, ArrayView<float> aData, ArrayView<float> aResult) {

            float nMax = 0;

            for (int i = 0; i < rows; i++) {
                float NP = aData[i * sample];
                float sum_xi = NP;
                float sum_ixi = 0;
                float M2 = 0;
                float sum_xi_save = sum_xi;

                for (int n = 1; n < sample; n++) {
                    NP += aData[i * sample + n];
                    sum_xi += NP;
                    float tmp = NP * n;
                    sum_ixi += tmp;
                    M2 += (tmp - sum_xi_save) * (tmp + NP - sum_xi) * aOpt[n];
                    sum_xi_save = sum_xi;
                }

                float npcod = (sum_ixi - (sample - 1) * sum_xi / 2);
                npcod *= npcod / (sample * M2 * (sample * sample - 1) / 12) * NP;

                nMax = (npcod + nMax + IntrinsicMath.Abs(npcod - nMax)) / 2;
            }

            aResult[nIndex] = nMax;
        }
    }
}

```

### Environment

- ILGPU version: 1.5.1
- .NET version: .NET 8
- Operating system: Windows 11
- Hardware : Intel(R) Iris(R) Xe Graphics

**Comment by MoFtZ** [comment]
@fcogneaux I haven't been able to reproduce the issue.

However, that `CLException` that is thrown should have an `Error` member variable that will give you the error code from the OpenCL API. In particular, it should map to one of [these error codes](https://github.com/m4rs-mt/ILGPU/blob/18ed8ee9e4fe45f76a7176963b1b9b051eb0ac4d/Src/ILGPU/Runtime/OpenCL/CLNativeMethods.cs#L19).

The call to `Synchronize` will internally call the `clFinish` function of OpenCL. [The documentation](https://registry.khronos.org/OpenCL/sdk/2.0/docs/man/xhtml/) says that [clFinish](https://registry.khronos.org/OpenCL/sdk/2.0/docs/man/xhtml/clFinish.html) can expect the following result:
```
 Returns CL_SUCCESS if the function call was executed successfully. Otherwise, it returns one of the following errors:

    CL_INVALID_COMMAND_QUEUE if command_queue is not a valid host command-queue.
    CL_OUT_OF_RESOURCES if there is a failure to allocate resources required by the OpenCL implementation on the device.
    CL_OUT_OF_HOST_MEMORY if there is a failure to allocate resources required by the OpenCL implementation on the host.
```

Let us know how you go.

**Comment by fcogneaux** [comment]
Thank you for your time and you suggestion MoFtZ.

Indeed, the exception has an error code which is : -14 (CL_EXEC_STATUS_ERROR_FOR_EVENTS_IN_WAIT_LIST)

Given the context (a low powered graphic card), and accordingly to the thread posts I was able to find, it could be related to some watchdog / timeout error, raised by the hardware when a kernel takes too much time to execute.

I found a workaround by splitting the outer loop in smaller chunks before execute the kernel.

Thanks!
